server/src/main/java/io/druid/server/QueryResource.java

```diff
@@ -100,16 +111,39 @@ public QueryResource(
     this.emitter = emitter;
     this.requestLogger = requestLogger;
     this.queryManager = queryManager;
+    this.authConfig = authConfig;
   }
 
   @DELETE
   @Path("{id}")
   @Produces(MediaType.APPLICATION_JSON)
-  public Response getServer(@PathParam("id") String queryId)
+  public Response getServer(@PathParam("id") String queryId, @Context final HttpServletRequest req)
   {
+    if (authConfig.isEnabled()) {
+      // This is an experimental feature, see - https://github.com/druid-io/druid/pull/2424
+      final AuthorizationInfo authorizationInfo = (AuthorizationInfo) req.getAttribute(AuthConfig.DRUID_AUTH_TOKEN);
+      Preconditions.checkNotNull(
+          authorizationInfo,
+          "Security is enabled but no authorization info found in the request"
+      );
+      if (queryManager.getQueryDatasources(queryId) == null) {
+        log.warn("QueryId [%s] not registered with QueryManager, cannot cancel", queryId);
+      } else if (authorizationInfo != null) {
+        for (String dataSource : queryManager.getQueryDatasources(queryId)) {
```

Comment:
race condition. Should store result of `queryManager.getQueryDatasources(queryId)` and not assume the method will return the same result each call.